### PR TITLE
Cancelling of duplicate tasks due to entity persistence

### LINKED ISSUE #734: Implement cancelling of duplicate tasks on Android
- [ ] Due to the implementation of Entity persistence, locally created tasks should be transfer data to server created tasks and then be deleted so that they are not synced later.

- [ ] We need to first sync tasks from server before syncing tasks created on the device.

Whenever there is a task synced from the server and local task exists and is not in Ready status, the below should happen

- [ ] The local task will be used to update the task synced from the server
- [ ] The events will be updated to reference the new task
- [ ] Client processing should be triggered on events above
- [ ] The local task will then be deleted
- [ ] The task and event will be synced to the server

### PR DESCRIPTION
**Pull request by Rkareko**
Resolves #734

### CODE REVIEW

File: opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java [hunk 1]
```diff
@@ -2258,4 +2271,12 @@ public boolean deleteEventsByBaseEntityId(String baseEntityId, String eventType)
         public String sqlQuery;
         public Map<String, Integer> columnOrder;
     }
+
+    public List<Event> getEventsByTaskIds(Set<String> taskIds) {
+        return fetchEvents("SELECT json FROM "
+                + eventTable.name()
+                + " WHERE "
+                + event_column.taskId.name()
+                + " IN (" + StringUtils.repeat(",", taskIds.size()) + ")", taskIds.toArray(new String[0]));
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/repository/EventClientRepository.java:2280]:
```suggestion
                + " IN (" + StringUtils.repeat("?",",", taskIds.size()) + ")", taskIds.toArray(new String[0]));
```

File: opensrp-app/src/main/java/org/smartregister/repository/TaskRepository.java [hunk 2]
```diff
@@ -684,4 +684,48 @@ public int getUnsyncedCreatedTasksAndTaskStatusCount() {
...
+        List<String> entityIds = new ArrayList<>();
+        android.database.Cursor cursor = null;
+        String query = "SELECT\n" +
+                "    DISTINCT(for)\n" +
+                "FROM\n" +
+                "    task\n" +
+                "GROUP BY\n" +
+                "    plan_id, for, code\n" +
+                "HAVING \n" +
+                "    COUNT(*) > 1";
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/repository/TaskRepository.java:698]: can we remove the \n, maybe we use String.format?

File: opensrp-app/src/main/java/org/smartregister/repository/TaskRepository.java [hunk 3]
```diff
@@ -684,4 +684,48 @@ public int getUnsyncedCreatedTasksAndTaskStatusCount() {
...
+                "    task\n" +
+                "GROUP BY\n" +
+                "    plan_id, for, code\n" +
+                "HAVING \n" +
+                "    COUNT(*) > 1) t2\n" +
+                "ON t1.plan_id = t2.plan_id\n" +
+                "AND t1.for = t2.for\n" +
+                "AND t1.code = t2.code\n" +
+                "AND t1.for = ? \n" +
+                "ORDER BY t1.for";
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/repository/TaskRepository.java:723]: can we remove \n for readability of the query, you can use space at the end/start of each line

File: opensrp-app/src/main/java/org/smartregister/sync/helper/TaskServiceProcessor.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        if (duplicateTaskList.get(0).getServerVersion() > duplicateTaskList.get(1).getServerVersion()){
+            serverTask = duplicateTaskList.get(0);
+            localTask = duplicateTaskList.get(1);
+        } else {
+            serverTask = duplicateTaskList.get(1);
+            localTask = duplicateTaskList.get(0);
+        }
+
+        if (Task.TaskStatus.READY.equals(localTask.getStatus())){
+            return;
```
Comment (githengi) [inline opensrp-app/src/main/java/org/smartregister/sync/helper/TaskServiceProcessor.java:75]: if the local task if ready maybe the population should be skipped, however the task still needs to be deleted

**Review by githengi — CHANGES_REQUESTED** [review]

File: opensrp-app/src/main/java/org/smartregister/repository/TaskRepository.java [hunk 5]
```diff
@@ -684,4 +684,48 @@ public int getUnsyncedCreatedTasksAndTaskStatusCount() {
...
+            entityIds.add(cursor.getString(0));
+        }
+        return entityIds;
+    }
+
+    public Set<Task> getDuplicateTasksForEntity(String entityId) {
+        String query = "SELECT t1.* " +
+                "FROM task t1 " +
+                "JOIN (SELECT " +
+                "    _id, plan_id, for, code, COUNT(*) as count " +
```
[the remark this answers is not in the document]
Comment (Rkareko) [reply opensrp-app/src/main/java/org/smartregister/repository/TaskRepository.java]: removed

**Review by githengi — APPROVED** [review]

**Pull request merged**